api/common/sharing/utils.go

```diff
@@ -12,44 +12,223 @@ import (
+	plan, planExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	labels, labelsExists := parameters[instance_sharing.ReferenceLabelSelector]
+	var selectorLabels types.Labels
+	if err := util.BytesToObject([]byte(labels.Raw), &selectorLabels); labelsExists && err != nil {
+		return err
 	}
+	hasAtLeastOneSelector := selectorLabels != nil && len(selectorLabels) > 0 || nameExists && len(name.String()) > 0 || planExists && len(plan.String()) > 0
+	selectorsAndID := IDExists && len(ID.String()) > 0 && hasAtLeastOneSelector
+	emptyValues := len(ID.String()) == 0 && !hasAtLeastOneSelector
+	if selectorsAndID || emptyValues {
+		return util.HandleInstanceSharingError(util.ErrInvalidReferenceSelectors, "")
+	}
+	return nil
+}
+
+func getSharedInstances(ctx context.Context, repository storage.Repository, offeringID, tenantIdentifier, tenantID string) (types.ObjectList, error) {
+	queryParams := map[string]interface{}{
+		"tenant_identifier": tenantIdentifier,
+		"tenant_id":         tenantID,
+		"offering_id":       offeringID,
+	}
+	objectList, err := repository.QueryForList(ctx, types.ServiceInstanceType, storage.QueryForSharedInstances, queryParams)
+	return objectList, err
+}
+
+func filterInstancesBySelectors(ctx context.Context, repository storage.Repository, instances types.ObjectList, parameters map[string]gjson.Result, smaap bool, offeringID string) (*types.ServiceInstance, error) {
+	var err error
+	var filteredInstances types.ServiceInstances
+
+	planSelectorVal, planSelectorExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	var selectorPlan *types.ServicePlan
+	if planSelectorExists && len(planSelectorVal.String()) > 0 {
+		selectorPlan, err = getSelectorPlan(ctx, repository, smaap, offeringID, planSelectorVal.String())
+		if err != nil {
+			log.C(ctx).Errorf("Failed to retrieve the plan %s with the offering %s via the selector provided: %s", planSelectorVal.String(), offeringID, err)
+			return nil, err
+		}
+	}
+
+	for i := 0; i < instances.Len(); i++ {
+		// "*" for any shared instance of the tenant, if parameter not passed, set true for any shared instance
+		selectorVal, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+		if exists && len(selectorVal.String()) > 0 && selectorVal.String() != "*" {
+			continue
+		}
+
+		instance := instances.ItemAt(i).(*types.ServiceInstance)
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceInstanceNameSelector]
+		if exists && len(selectorVal.String()) > 0 && !(instance.Name == selectorVal.String()) {
+			continue
+		}
+
+		if selectorPlan != nil && instance.ServicePlanID != selectorPlan.ID {
+			continue
+		}
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceLabelSelector]
+		if exists {
+			var selectorLabels types.Labels
+			if err := util.BytesToObject([]byte(selectorVal.Raw), &selectorLabels); err != nil {
+				log.C(ctx).Errorf("%s", err)
+				return nil, err
```

Comment:
if you have nothing to add don't log the error it will be logged later
you have more than one

Reply:
i will lower the amount of logs.